packages/react-meteor-data/useTracker.js

```diff
@@ -0,0 +1,155 @@
+/* global Meteor, Package, Tracker */
+import React, { useState, useEffect, useRef } from 'react';
+
+// Use React.warn() if available (should ship in React 16.9).
+const warn = React.warn || console.warn.bind(console);
+
+// Warns if data is a Mongo.Cursor or a POJO containing a Mongo.Cursor.
+function checkCursor(data) {
+  let shouldWarn = false;
+  if (Package.mongo && Package.mongo.Mongo && data && typeof data === 'object') {
+    if (data instanceof Package.mongo.Mongo.Cursor) {
+      shouldWarn = true;
+    } else if (Object.getPrototypeOf(data) === Object.prototype) {
+      Object.keys(data).forEach((key) => {
+        if (data[key] instanceof Package.mongo.Mongo.Cursor) {
+          shouldWarn = true;
+        }
+      });
+    }
+  }
+  if (shouldWarn) {
+    warn(
+      'Warning: your reactive function is returning a Mongo cursor. '
+      + 'This value will not be reactive. You probably want to call '
+      + '`.fetch()` on the cursor before returning it.'
+    );
+  }
+}
+
+// taken from https://github.com/facebook/react/blob/
+// 34ce57ae751e0952fd12ab532a3e5694445897ea/packages/shared/objectIs.js
+function is(x, y) {
+  return (
+    (x === y && (x !== 0 || 1 / x === 1 / y))
+    || (x !== x && y !== y) // eslint-disable-line no-self-compare
+  );
+}
+
+// inspired by https://github.com/facebook/react/blob/
+// 34ce57ae751e0952fd12ab532a3e5694445897ea/packages/
+// react-reconciler/src/ReactFiberHooks.js#L307-L354
+// used to replicate dep change behavior and stay consistent
+// with React.useEffect()
+function areHookInputsEqual(nextDeps, prevDeps) {
+  if (prevDeps === null || prevDeps === undefined || !Array.isArray(prevDeps)) {
+    return false;
+  }
+
+  if (!Array.isArray(nextDeps)) {
```

Comment:
Should this also check if `nextDeps` is falsy? If `nextDeps` is passed in as `null`, `undefined`, or `false` this sill trigger a positive match, but I think that's not correct at least for the warning (the return value seem correct).

Reply:
This was taken from https://github.com/meteor/react-packages/blob/bff72c7c291743acd5348f81846ea2679000666c/packages/react-meteor-data/useTracker.js#L49

```JS
  if (!Array.isArray(nextDeps)) {
    if (Meteor.isDevelopment) {
      warn(
        'Warning: useTracker expected an dependency value of '
        + `type array but got type of ${typeof nextDeps} instead.`
      );
    }
    return false;
  }
```

I am not sure if I understand you correctly here, this should take care of all scenarios?:

```JS
!Array.isArray(undefined) => true
!Array.isArray(null) => true
!Array.isArray(0) => true
!Array.isArray([]) => false
```

Reply:
Oh, you are right, this is fine.

Reply:
@CaptainN On second thought, I think react is doing that only for performance reasons. Maybe early exiting the check on null or undefined is faster than Array.isArray(), I don't mind if you change it.

Reply:
You might be interested in https://github.com/risetechnologies/react-packages/commit/e3bc3656586b812728ae57fe91222e48cd00ad66 `areDepsValid`, but be careful this is WIP

Reply:
I think it all matches now, does it not?

I would have thought a falsy check would be enough (`if (!deps && !Array.isArray(deps)) {`) - but the React code seems to specifically avoid checking `false`.

Reply:
I am not really sure (working on this code every now and then is really confusing, always losing something in between). Right now I read: If the nextDeps is not an array, always warn. But nextDeps can actually be null or undefined per spec, so when in development it warns.

I think we might not need a change here because since prevDeps and nextDeps usually should always be the same type, e.g. once deps are null, they should stay null. Since the prevDeps check happens before and early returns false, this code *should* be able to stay like it is.

Reply:
I actually think it should treat changes to deps as a fresh invokation every time (the react team gets this wrong in their code comment). Your `react-router` `Switch` example demonstrates why. Sometimes the React engine will re-use an instance in ways you don't expect - in those cases, it really is like starting over. I've kept that in mind all the way through the development of this, and I think it's lead to a rock solid code base. Here I think the issue is actually relatively minor - but maybe the fix is too:

```js
  if (!Array.isArray(nextDeps)) {
    if (Meteor.isDevelopment && deps !== undefined && deps !== null ) {
      warn(
        'Warning: useTracker expected an dependency value of '
        + `type array but got type of ${typeof nextDeps} instead.`
      );
    }
    return false;
  }
```

Reply:
looks perfect! 🚀